providers/implementations/signature/ml_dsa_sig.c

```diff
@@ -0,0 +1,245 @@
+static OSSL_FUNC_signature_settable_ctx_params_fn ml_dsa_settable_ctx_params;
+
+typedef struct {
+    ML_DSA_KEY *key;
+    ML_DSA_CTX *ctx;
+    uint8_t context_string[ML_DSA_MAX_CONTEXT_STRING_LEN];
+    size_t context_string_len;
+    uint8_t test_entropy[ML_DSA_ENTROPY_LEN];
+    size_t test_entropy_len;
+    int msg_encode;
+    int deterministic;
+    OSSL_LIB_CTX *libctx;
+    char *propq;
+} PROV_ML_DSA_CTX;
+
+static void ml_dsa_freectx(void *vctx)
+{
+    PROV_ML_DSA_CTX *ctx = (PROV_ML_DSA_CTX *)vctx;
+
+    OPENSSL_free(ctx->propq);
+    ossl_ml_dsa_ctx_free(ctx->ctx);
+    OPENSSL_cleanse(ctx->test_entropy, ctx->test_entropy_len);
+    OPENSSL_free(ctx);
+}
+
+static void *ml_dsa_newctx(void *provctx, const char *alg, const char *propq)
+{
+    PROV_ML_DSA_CTX *ctx;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    ctx = OPENSSL_zalloc(sizeof(PROV_ML_DSA_CTX));
+    if (ctx == NULL)
+        return NULL;
+
+    ctx->libctx = PROV_LIBCTX_OF(provctx);
+    if (propq != NULL && (ctx->propq = OPENSSL_strdup(propq)) == NULL)
+        goto err;
+    ctx->ctx = ossl_ml_dsa_ctx_new(alg, ctx->libctx, ctx->propq);
+    if (ctx->ctx == NULL)
+        goto err;
+    ctx->msg_encode = ML_DSA_MESSAGE_ENCODE_PURE;
+
+    return ctx;
+ err:
+    ml_dsa_freectx(ctx);
+    return NULL;
+}
+
+static int ml_dsa_signverify_msg_init(void *vctx, void *vkey,
+                                      const OSSL_PARAM params[], int operation,
+                                      const char *desc)
+{
+    PROV_ML_DSA_CTX *ctx = (PROV_ML_DSA_CTX *)vctx;
+    ML_DSA_KEY *key = vkey;
+
+    if (!ossl_prov_is_running()
+            || ctx == NULL)
+        return 0;
+
+    if (vkey == NULL && ctx->key == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_KEY_SET);
+        return 0;
+    }
+
+    if (key != NULL) {
+        if (!ossl_ml_dsa_key_type_matches(ctx->ctx, key))
+            return 0;
+        ctx->key = vkey;
+    }
+
+    if (!ml_dsa_set_ctx_params(ctx, params))
+        return 0;
+    return 1;
+}
+
+static int ml_dsa_sign_msg_init(void *vctx, void *vkey, const OSSL_PARAM params[])
+{
+    return ml_dsa_signverify_msg_init(vctx, vkey, params,
+                                      EVP_PKEY_OP_SIGN, "ML_DSA Sign Init");
+}
+
+static int ml_dsa_sign(void *vctx, unsigned char *sig, size_t *siglen,
+                       size_t sigsize, const unsigned char *msg, size_t msg_len)
+{
+    int ret = 0;
+    PROV_ML_DSA_CTX *ctx = (PROV_ML_DSA_CTX *)vctx;
+    uint8_t rand_tmp[ML_DSA_ENTROPY_LEN], *rnd = NULL;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (sig != NULL) {
+        if (ctx->test_entropy_len != 0) {
+            rnd = ctx->test_entropy;
+        } else {
+            rnd = rand_tmp;
+
+            if (ctx->deterministic == 1)
+                memset(rnd, 0, sizeof(rand_tmp));
+            else if (RAND_priv_bytes_ex(ctx->libctx, rnd, sizeof(rand_tmp), 0) <= 0)
+                return 0;
```

Comment:
Might want an ERR_raise here - unless RAND_priv_bytes_ex is going to do that before returning ...

Reply:
In the places that matter (such as generate) this raises an error already.